python/gps/algorithm/algorithm_traj_opt_pilqr.py

```diff
@@ -0,0 +1,166 @@
+            self.cur[m].step_mult = self.prev[m].step_mult
+        elif self._hyperparams['step_rule'] == 'res_percent':
+            cur_res_sum = np.mean(np.sum(np.abs(cur_res), axis=1))
+            act_sum = np.mean(np.sum(np.abs(self.cur[m].cs), axis=1))
+            # Print cost details.
+            LOGGER.debug(
+                    'Cur residuals: %f. Residuals percentage: %f',
+                    cur_res_sum, cur_res_sum/act_sum*100
+            )
+            if self.traj_opt.cons_per_step:
+                T = cur_res.shape[1]
+                new_mult = np.ones(T)
+                for t in range(T):
+                    res = np.mean(np.sum(np.abs(cur_res[:, t:]), axis=1))
+                    act = np.mean(np.sum(np.abs(self.cur[m].cs[:, t:]), axis=1))
+                    if res/act > self._hyperparams['step_rule_res_ratio_dec']:
+                        new_mult[t] = 0.5
+                    elif res/act < self._hyperparams['step_rule_res_ratio_inc']:
+                        new_mult[t] = 2.0
+            else:
+                if cur_res_sum/act_sum > self._hyperparams['step_rule_res_ratio_dec']:
+                    new_mult = 0.5
+                elif cur_res_sum/act_sum < self._hyperparams['step_rule_res_ratio_inc']:
+                    new_mult = 2.0
+                else:
+                    new_mult = 1
+            new_step = np.maximum(
+                    np.minimum(new_mult * self.cur[m].step_mult, max_mult),
+                    min_mult
+            )
+            self.cur[m].step_mult = new_step
+        else:
+            raise NotImplementedError('Unknown step adjustment rule.')
+
+    def _compute_costs(self, traj_distr, m, eta, augment=True):
+        """ Compute cost estimates used in the LQR backward pass. """
+
+        traj_info, traj_distr = self.cur[m].traj_info, traj_distr
+        if not augment:
+            return traj_info.Cm, traj_info.cv
+
+        multiplier = self._hyperparams['max_ent_traj']
+        fCm, fcv = traj_info.Cm / (eta + multiplier), traj_info.cv / (eta + multiplier)
+        K, ipc, k = traj_distr.K, traj_distr.inv_pol_covar, traj_distr.k
+
+        # Add in the trajectory divergence term.
+        for t in range(self.T - 1, -1, -1):
+            fCm[t, :, :] += eta / (eta + multiplier) * np.vstack([
+                np.hstack([
+                    K[t, :, :].T.dot(ipc[t, :, :]).dot(K[t, :, :]),
+                    -K[t, :, :].T.dot(ipc[t, :, :])
+                ]),
+                np.hstack([
+                    -ipc[t, :, :].dot(K[t, :, :]), ipc[t, :, :]
+                ])
+            ])
+            fcv[t, :] += eta / (eta + multiplier) * np.hstack([
+                K[t, :, :].T.dot(ipc[t, :, :]).dot(k[t, :]),
+                -ipc[t, :, :].dot(k[t, :])
+            ])
+
+        return fCm, fcv
+
+    def compute_costs(self, m, eta, augment=True):
+        return self._compute_costs(self.cur[m].traj_distr, m, eta, augment)
```

Comment:
Why not just have a single function that sets `traj_distr = self.cur[m].traj_distr` at the top?